### PR TITLE
IM-203 - Allow Admins To Switch Payment Gateway Type to Merchant ID

### PR DESCRIPTION
**Pull request by dougworner**
Allow Admins To Switch Payment Gateway Type to Merchant ID

Currently, if you create a conference, you only have the option to add a merchant id for Payment Options Credit Card payment. This is what we want.

However, if you already have a conference that takes Authorize.NET credentials, we'd like the conference admin to be able to switch to Merchant ID.

Add to the Payment Options Credit Card section a Payment Gateway Type menu with two options, 'Authorize.NET' and 'TSYS Merchant ID', but only if the current payment gateway type is 'Authorize.NET'. Otherwise, the conference must stay as Merchant ID.

### CODE REVIEW

File: test/spec/controllers/eventDetails.js [hunk 1]
```diff
@@ -1,65 +0,0 @@
-'use strict';
```
Comment (canac) [inline test/spec/controllers/eventDetails.js]: Why are you deleting this file? Are the tests stale?

File: app/views/eventDetails/paymentOptions.html [hunk 2]
```diff
@@ -5,6 +5,15 @@ <h2 class="page-title border" translate>Payment Options</h2>
     <h3 class="panel-title" translate>Credit Card</h3>
   </div>
   <div class="panel-body">
+    <div class="form-group">
+      <div class="col-sm-12 stacked-spacing-col-sm" ng-if="conference.paymentGatewayType === 'AUTHORIZE_NET'">
+        <label translate>Select A Payment Gateway Type:</label>
+        <select class="form-control" ng-model="conference.paymentGatewayType">
+          <option value="TSYS" translate>TSYS Merchant ID</option>
```
Comment (canac) [inline app/views/eventDetails/paymentOptions.html]: I think you mean TSYS here.

File: app/views/eventDetails/paymentOptions.html [hunk 3]
```diff
@@ -35,6 +44,7 @@ <h3 class="panel-title" translate>Credit Card</h3>
         </div>
       </div>
     </div>
+
```
Comment (canac) [inline app/views/eventDetails/paymentOptions.html]: Can you remove this stray line-break?

File: app/views/eventDetails/paymentOptions.html [hunk 4]
```diff
@@ -5,6 +5,15 @@ <h2 class="page-title border" translate>Payment Options</h2>
     <h3 class="panel-title" translate>Credit Card</h3>
   </div>
   <div class="panel-body">
+    <div class="form-group">
+      <div class="col-sm-12 stacked-spacing-col-sm" ng-if="conference.paymentGatewayType === 'AUTHORIZE_NET'">
+        <label translate>Select A Payment Gateway Type:</label>
+        <select class="form-control" ng-model="conference.paymentGatewayType">
```
Comment (canac) [inline app/views/eventDetails/paymentOptions.html]: These options could be generated by the data in `$scope.paymentGateways`. I won't say that you have to, but it would reduce duplication of the labels. You would use `ng-options` to specify an array of options generated from `$scope.paymentGateways`. There would be one option per entry, with the key being the option value and the `name` field being option label. Let me know if you decide to go that route and need help figuring it out.

File: app/views/eventDetails/paymentOptions.html [hunk 5]
```diff
@@ -61,57 +71,58 @@ <h3 class="panel-title" translate>Checks</h3>
       </div>
       <div class="col-sm-3 stacked-spacing-col-sm">
         <select class="form-control" ng-model="conference.checkMailingState">
-          <option value="AL">Alabama</option>
```
Comment (canac) [inline app/views/eventDetails/paymentOptions.html]: Great job catching these untranslated labels! 👍

File: test/spec/controllers/eventDetails.js (hunk 1 above)
Comment (dougworner) [inline test/spec/controllers/eventDetails.js]: I was having issues with Intellij and debugging. instead of using the actual eventDetails.js file when setting breakpoints it was the test/spec/controllers/eventDetails.js file.  So if I made a code change and then refreshed browser the breakpoints would not be accurate.  So I was forced to stop, save, restart every time I made a change which took longer. If there is a better approach please let me know.

File: app/views/eventDetails/paymentOptions.html (hunk 5 above)
Comment (dougworner) [inline app/views/eventDetails/paymentOptions.html]: Thanks.  I did some research to make sure we could use translate with and select option with gettext but really could not find a lot of documentation with this specific use case.  But also didn't find anything against.

File: app/views/eventDetails/paymentOptions.html (hunk 2 above)
Comment (dougworner) [inline app/views/eventDetails/paymentOptions.html]: I was going strictly off of what IM-203 Jira says, "two options, 'Authorize.NET' and 'TSYS Merchant ID'."
So I used 'Authorize.NET' and 'TSYS Merchange ID'.
Do you think it would be better for end users to make this select option 'TSYS'?

File: app/views/eventDetails/paymentOptions.html (hunk 3 above)
Comment (dougworner) [inline app/views/eventDetails/paymentOptions.html]: Think I got it.

File: app/views/eventDetails/paymentOptions.html (hunk 4 above)
Comment (dougworner) [inline app/views/eventDetails/paymentOptions.html]: I think that is a good suggestion (less code).  I made the change, but waiting to hear back from Lee as the link I used to use at the bottom of the https://stage.eventregistrationtool.com/#/ site is now missing (i.e. ‘Local Dev Environment’).
I would always click on this link after login to Cru Relay to do dev/test locally.
Waiting to push changes until this is resolved.

File: app/views/eventDetails/paymentOptions.html (hunk 2 above)
Comment (dougworner) [inline app/views/eventDetails/paymentOptions.html]: if going with the ng-options it will end up being 'TSYS' since will be using the name field.

**Comment by hlbraddock** [comment]
@dougworner Can you incorporate changes as per @canac 's comments? Thanks!

File: test/spec/controllers/eventDetails.js (hunk 1 above)
Comment (canac) [inline test/spec/controllers/eventDetails.js]: It's fine if you keep the file deleted locally, just don't commit that deletion. Please revert this change. If you need help with the git commands, let me know.

File: app/views/eventDetails/paymentOptions.html (hunk 5 above)
Comment (canac) [inline app/views/eventDetails/paymentOptions.html]: From the other places it is used, it looks like putting the translate attribute on an element translates the text content of that element. Since it is the text content of the option elements that we want to translate, I think you are good to do it that way.

File: app/views/eventDetails/paymentOptions.html (hunk 3 above)
Comment (canac) [inline app/views/eventDetails/paymentOptions.html]: 👍

File: app/views/eventDetails/paymentOptions.html (hunk 2 above)
Comment (canac) [inline app/views/eventDetails/paymentOptions.html]: Yes, I do. A payment gateway of "TSYS Merchant ID" makes no sense to me. We can check with @hlbraddock, but I bet it was just a typo in the JIRA ticket.

**Comment by dougworner** [comment]
@hlbraddock thanks for catching and "removing display of gateway type".  Test code I left in.

File: test/spec/controllers/eventDetails.js (hunk 1 above)
Comment (dougworner) [inline test/spec/controllers/eventDetails.js]: This happened before I completely recreated local environment with Git history of deleting this file.  So I recreated the file and pushed branch.

File: app/views/eventDetails/paymentOptions.html (hunk 2 above)
Comment (dougworner) [inline app/views/eventDetails/paymentOptions.html]: I agree.  We can double check with @hlbraddock if you think it is needed.

**Comment by canac** [comment]
Here's my suggestion. Save a reference to `conference` in `$scope.originalConference`. Also, whenever the conference is successfully saved, update `$scope.originalConference`. Now `$scope.originalConference` will always refer to the conference data as stored on the server, and `$scope.conference` will refer to the conference with any unsaved changes. Then in `paymentOptions.html`, we can update the `ng-if` condition to only show the payment gateway type selector when `$scope.originalConference === 'AUTHORIZE_NET'`.

**Comment by canac** [comment]
@dougworner Lee wanted the change that I suggested in my comment above implemented quickly, so I went ahead and made it.

**Comment by hlbraddock** [comment]
Great work @dougworner !

**Comment by dougworner** [comment]
Thank you @hlbraddock .
Sorry had to run to bank this morning otherwise I would have implemented Caleb's change.
So thankful to be able to help you out!

**Comment by dougworner** [comment]
@canac Thanks for making the last change. Your suggestion makes sense and be one I remember for next time.

**Comment by dougworner** [comment]
@canac Question to better understand...is the reason $scope.conference represents unsaved data because this object is not updated upon clicking the save button?  But if a person did a screen refresh it would be updated?

**Comment by canac** [comment]
@dougworner `$scope.conference` reflects whatever data is displayed in the forms because it is bound to the form elements via `ng-model`. If the user changes the conference name, `$scope.conference.name` will contain the new name, but the old name will still be in the database because it has not been saved yet. `$scope.originalConference.name` will match the database value: the old conference name. If the user did a page refresh before saving, changes to `$scope.conference` will be lost. Does that answer your question?

**Comment by dougworner** [comment]
@canac Yah, that makes it more clear.  Because $scope.conference only exists in current memory but not in the database (i.e. because save button has not been clicked to push these changes to the database), then any change you would make on the page would be immediately updated.  Only after clicking the save button would changes in $scope.conference be pushed to database.  So menu option will continue to display until save button is clicked and $scope.conference has been updated.  So not updated upon screen refresh but upon clicking saved button.
